File: csharp/DVSClient/DVSClient/address/format/AddressMetadata.cs
using DVSClient.Server.Address.Format;

namespace DVSClient.Address.Format
{
    public class AddressMetadata
    {
        public IEnumerable<string> Sources { get; }
        public string NumberOfHouseholds { get; }
        public string JustBuiltDate { get; }
        public string Umrrn { get; }
        public string Udprn { get; }
        public string Dpid { get; }
        public string Gnafpid { get; }
        public string PafAddressKey { get; }
        public string Hin { get; }
        public string DeliveryPointBarcode { get; }
        public string BarcodeCheckDigit { get; }
        public string BarcodeSortPlanNumber { get; }

        public AddressMetadata(RestApiAddressFormatMetadata apiMetadata)
        {
            var apiInfo = apiMetadata.AddressInfo;
            if (apiInfo != null)
            {
                Sources = apiInfo.Sources ?? new List<string>();
                NumberOfHouseholds = apiInfo.NumberOfHouseholds ?? string.Empty;
                JustBuiltDate = apiInfo.JustBuiltDate ?? string.Empty;

                var identifier = apiInfo.Identifier;
                if (identifier != null)
                {
                    Umrrn = identifier.Umrrn ?? string.Empty;
                    Udprn = identifier.Udprn ?? string.Empty;
                    Dpid = identifier.Dpid ?? string.Empty;
                    Gnafpid = identifier.Gnafpid ?? string.Empty;
                    PafAddressKey = identifier.PafAddressKey ?? string.Empty;
                    Hin = identifier.Hin ?? string.Empty;
                }
                else
                {
                    Umrrn = string.Empty;
                    Udprn = string.Empty;
                    Dpid = string.Empty;
                    Gnafpid = string.Empty;
                    PafAddressKey = string.Empty;
                    Hin = string.Empty;
                }
            }
            else
            {
                Sources = new List<string>();
                NumberOfHouseholds = string.Empty;
                JustBuiltDate = string.Empty;
                Umrrn = string.Empty;
                Udprn = string.Empty;
                Dpid = string.Empty;
                Gnafpid = string.Empty;
                PafAddressKey = string.Empty;
                Hin = string.Empty;
            }

            var barcode = apiMetadata.Barcode;
            if (barcode != null)
            {
                DeliveryPointBarcode = barcode.DeliveryPointBarcode ?? string.Empty;
                BarcodeCheckDigit = barcode.CheckDigit ?? string.Empty;
                BarcodeSortPlanNumber = barcode.SortPlanNumber ?? string.Empty;
            }
            else
            {
                DeliveryPointBarcode = string.Empty;
                BarcodeCheckDigit = string.Empty;
                BarcodeSortPlanNumber = string.Empty;
            }
        }
    }
}

File: typescript/src/address/format/addressMetadata.ts
import { RestApiAddressFormatMetadata } from "../../server/address/format/restApiAddressFormatMetadata";

export type AddressMetadata = {
    sources: string[];
    numberOfHouseholds: string;
    justBuiltDate: string;
    umrrn: string;
    udprn: string;
    dpid: string;
    gnafpid: string;
    pafAddressKey: string;
    hin: string;
    deliveryPointBarcode: string;
    barcodeCheckDigit: string;
    barcodeSortPlanNumber: string;
};

export function restApiResponseToAddressMetadata(response: RestApiAddressFormatMetadata): AddressMetadata {
    const apiInfo = response.address_info;
    const barcode = extractBarcode(response);

    if (apiInfo) {
        const identifier = apiInfo.identifier;
        return {
            sources: apiInfo.sources ?? [],
            numberOfHouseholds: apiInfo.number_of_households ?? "",
            justBuiltDate: apiInfo.just_built_date ?? "",
            ...extractIdentifier(identifier),
            ...barcode,
        };
    }

    return {
        sources: [],
        numberOfHouseholds: "",
        justBuiltDate: "",
        umrrn: "",
        udprn: "",
        dpid: "",
        gnafpid: "",
        pafAddressKey: "",
        hin: "",
        ...barcode,
    };
}

function extractIdentifier(identifier: any) {
    return {
        umrrn: identifier?.umrrn ?? "",
        udprn: identifier?.udprn ?? "",
        dpid: identifier?.dpid ?? "",
        gnafpid: identifier?.gnafpid ?? "",
        pafAddressKey: identifier?.paf_address_key ?? "",
        hin: identifier?.hin ?? "",
    };
}

function extractBarcode(response: RestApiAddressFormatMetadata) {
    return {
        deliveryPointBarcode: response.barcode?.delivery_point_barcode ?? "",
        barcodeCheckDigit: response.barcode?.check_digit ?? "",
        barcodeSortPlanNumber: response.barcode?.sort_plan_number ?? "",
    };
}